src/gt4py/next/iterator/type_system/type_synthesizer.py

```diff
@@ -266,82 +276,160 @@ def apply_lift(
+    domain: it_ts.DomainType, input_: ts.TypeSpec
+) -> it_ts.IteratorType:
+    """
+    Converts a field operation input into an iterator type, preserving its dimensions and data type.
+    """
+    input_dims = _collect_and_check_dimensions(input_)
+    element_type: ts.DataType = type_info.apply_to_primitive_constituents(
+        type_info.extract_dtype, input_
+    )
 
     return it_ts.IteratorType(
-        position_dims=domain.dims, defined_dims=defined_dims, element_type=element_type
+        position_dims=domain.dims, defined_dims=input_dims, element_type=element_type
     )
 
 
+def _handle_sparse_fields(input_: ts.FieldType | ts.TupleType | tuple[ts.FieldType, ts.TupleType]):
+    """
+    Processes sparse field inputs by removing LOCAL dimensions and converting the data type to ListType.
+    """
+    if isinstance(input_, (tuple, ts.TupleType)):
+        return ts.TupleType(types=[_handle_sparse_fields(field) for field in input_])
+    elif isinstance(input_, ts.FieldType):
+        input_dims = _collect_and_check_dimensions(input_)
+        element_type: ts.DataType = type_info.apply_to_primitive_constituents(
+            type_info.extract_dtype, input_
+        )
+
+        defined_dims = []
+        neighbor_dim = None
+        for dim in input_dims:
+            if dim.kind == common.DimensionKind.LOCAL:
+                assert neighbor_dim is None
+                neighbor_dim = dim
+            else:
+                defined_dims.append(dim)
+        if neighbor_dim:
+            element_type = ts.ListType(element_type=element_type, offset_type=neighbor_dim)
+
+        return ts.FieldType(dims=defined_dims, dtype=element_type)
+    elif isinstance(input_, ts.ScalarType):
+        return input_
+    else:
+        raise TypeError(f"Unexpected field type: {type(input_)}")
+
+
 @_register_builtin_type_synthesizer
 def as_fieldop(
     stencil: TypeSynthesizer,
     domain: Optional[it_ts.DomainType] = None,
     *,
     offset_provider_type: common.OffsetProviderType,
 ) -> TypeSynthesizer:
-    # In case we don't have a domain argument to `as_fieldop` we can not infer the exact result
-    # type. In order to still allow some passes which don't need this information to run before the
-    # domain inference, we continue with a dummy domain. One example is the CollapseTuple pass
-    # which only needs information about the structure, e.g. how many tuple elements does this node
-    # have, but not the dimensions of a field.
-    # Note that it might appear as if using the TraceShift pass would allow us to deduce the return
-    # type of `as_fieldop` without a domain, but this is not the case, since we don't have
-    # information on the ordering of dimensions. In this example
-    #   `as_fieldop(it1, it2 -> deref(it1) + deref(it2))(i_field, j_field)`
-    # it is unclear if the result has dimension I, J or J, I.
-    if domain is None:
-        domain = it_ts.DomainType(dims="unknown")
+    def _resolve_shift(
```

Comment:
As discussed (a while ago) I believe the interface should be more like `_resolve_shift(dims: list[Dimension], shift_tuple: ...) -> list[Dimension]`. Maybe even rename the function. It would be good if it is easy to understand why `reversed(shift_tuple[::2])` happens here. Additionally this should be a free function.